File: 백준/Gold/7579. 앱/앱.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

class Main {
    static StringTokenizer st;
    static int[][] dp;
    static int[] memories, costs;
    static int n, m, totalCost;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        st = new StringTokenizer(br.readLine());
        n = nextInt();
        m = nextInt();

        st = new StringTokenizer(br.readLine());
        memories = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            memories[i] = nextInt();
        }

        st = new StringTokenizer(br.readLine());
        costs = new int[n + 1];
        totalCost = 0;

        for (int i = 1; i <= n; i++) {
            costs[i] = nextInt();
            totalCost += costs[i];
        }

        dp = new int[n + 1][totalCost + 1];

        for (int i = 1; i <= n; i++) {
            int memory = memories[i];
            int cost = costs[i];

            for (int j = 0; j <= totalCost; j++) {
                if (cost > j) {
                    dp[i][j] = dp[i - 1][j];
                } else {
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - cost] + memory);
                }
            }
        }

        int res = -1;

        for (int i = 0; i <= totalCost; i++) {
            if (dp[n][i] >= m) {
                res = i;
                break;
            }
        }

        System.out.println(res);
        br.close();
    }

    static int nextInt() {
        return Integer.parseInt(st.nextToken());
    }
}

File: 백준/Gold/7579. 앱/앱.kt
import java.util.*
import kotlin.math.max

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = StringTokenizer(readLine()).run { nextToken().toInt() to nextToken().toInt() }
    val mems = StringTokenizer(readLine()).run { Array(n) { nextToken().toInt() } }
    val costs = StringTokenizer(readLine()).run { Array(n) { nextToken().toInt() } }
    close()

    var res = costs.sum()
    val dp = Array(n + 1) { IntArray(res + 1) }
    for (i in 1..n) {
        val c = costs[i - 1]

        for (j in 1..dp[0].lastIndex) {
            dp[i][j] = if (j < c)
                dp[i - 1][j]
            else
                max(dp[i - 1][j - c] + mems[i - 1], dp[i - 1][j])

            if (dp[i][j] >= m && res > j) {
                res = j
            }
        }
    }
    print(res)
}